spices/0002-uri-rewriting-for-mirroring.adoc

```diff
@@ -0,0 +1,319 @@
+      -1
+    else if (rewrite.replace is Regex)
+      let (match = rewrite.replace.findMatchesIn(uri).first)
+        match.end - match.start
+    else
+      rewrite.replace.length
+  )
+
+  maxScore = matchesByScore.keys.max
+
+  rewritten: Uri = if (maxScore == -1) uri else
+    let (rewrite = matchesByScore[maxScore].first)
+      uri.replaceFirst(rewrite.replace, rewrite.with)
+}
+
+function rewrite(inputUri: Uri, uriRewrites: List<UriRewrite>): Uri = new Rewrite {
+  rewrites = uriRewrites
+  uri = inputUri
+}.rewritten
+----
+
+=== Implementation
+
+The proposed design is to implement this behaviour in `IoUtils::resolve`.
+`IoUtils::resolve` is used throughout to resolve URIs, regardless of whether a URI is used in an `import`, `import*`, `extends`, `amends`, `read`, or `read*`.
+
+=== Error messages
+
+Rewriting URIs according to definitions _not_ in the modules being evaluated could lead to hard to debug failures.
+When a rewritten URI points to an invalid resource, for example, users must be informed of the URIs provenance.
+The solution is to present the URI rewrite as a stack frame, identifying the applied rewrite rule.
+
+Rewriting is reported as follows
+[source]
+----
+Rewriting URI <URI_as_in_Pkl_code>
+  replacing "<string_or_regex_from_applied_rule>"
+  with "<replacement_string_from_applied_rule>"
+at <source_position_of_applied_rule>
+----
+
+As an example, consider a file `PklProject`, which defines
+[source,pkl]
+----
+amends "pkl:Project"
+
+uriRewrites {
+  new {
+    replace = "example.com"
+    with = "domain-that-does-not-really-exist.con"
+  }
+}
+----
+and a file `attempt.pkl` containing
+[source,pkl]
+----
+import "https://example.com/dependency.pkl"
+
+name = dependency.name
+----
+
+Evaluating this wants to resolve `import "https://example.com/dependency.pkl"`, which triggers the rewrite rule to produce the rewritten URI `https://domain-that-does-not-really-exist.con/dependency.pkl`.
+Since the rewritten URI points to a non-existent domain, Pkl produces an error:
+
+[source,bash]
+----
+$ pkl eval attempt.pkl
+–– Pkl Error ––
+I/O error loading module `https://domain-that-does-not-really-exist.con/dependency.pkl`.
+UnknownHostException: domain-that-does-not-really-exist.con
+
+Rewriting URI "https://example.com/dependency.pkl"
+  replacing "example.com"
+  with "domain-that-does-not-really-exist.con"
+at PklProject (file:///Users/jappleseed/demo/PklProject, line 4)
+
+1 | import "https://domain-that-does-not-really-exist.con/dependency.pkl"
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+at attempt#dependency (file:///Users/jappleseed/demo/attempt.pkl, line 1)
+
+3 | name = dependency.name
+           ^^^^^^^^^^
+at attempt#name (file:///Users/jappleseed/demo/attempt.pkl, line 3)
+
+106 | text = renderer.renderDocument(value)
+             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+at pkl.base#Module.output.text (https://github.com/apple/pkl/blob/0.25.2/stdlib/base.pkl#L106)
+----
```

Comment:
How would this work if the rewrite rule is provided via another means, like via a CLI flag?

Also, it doesn't sound great to insert stack frames to point to nodes that aren't actually being executed. We do this sparingly (e.g. to represent values that initialized as parse-time constants, and a language implementation optimization detail).

Rather than this, I suggest that we decorate the error with hints. Take a look at `VmExceptionBuilder#withHint`.